### PR TITLE
Run single instance of a pipeline without locking at end (unlock when done) (#106)

### PR DESCRIPTION
**Pull request by arvindsv**
The ability to unlock a pipeline automatically, upon failure. So, run a single instance of a pipeline run but don't lock if it fails. See #106.

Todos:
- [x] Func tests: Fix existing functional tests! (PR: https://github.com/gocd/functional-tests/pull/229)
- [ ] Func tests: Write a functional test for this behavior.
- [x] Docs: Change docs (both API docs and user docs). (PRs: https://github.com/gocd/api.go.cd/pull/213 and https://github.com/gocd/docs.go.cd/pull/122)
- [x] Docs: Update schema reference in docs. ([See diff in PR](https://github.com/gocd/docs.go.cd/pull/122/files#diff-a7b70224eea5fa58ad1c265463237ca5R1172))
- [x] Docs: Document missing plugin endpoint (config repo pipelines) (PR: https://github.com/gocd/plugin-api.go.cd/pull/44)
- [x] External: Update config repo plugins to provide version and ability to set this property. (PRs: https://github.com/tomzo/gocd-json-config-plugin/pull/12 and https://github.com/tomzo/gocd-yaml-config-plugin/pull/46)
- [x] External: Inform gomatic team about this change (@dudadornelles opened https://github.com/gocd-contrib/gomatic/issues/46)
- [x] Check: Make sure this works. :)
- [x] Check: Does canceling a build release the lock? Think it does.
- [x] Check: Errors in API look fine.
- [x] Check: Ensure API v3 and v4 of pipeline config continue to work.
- [x] Debt: Upgrade jolt to 0.1.0

This image is to help @jyotisingh decipher [this test](https://github.com/gocd/gocd/compare/master...arvindsv:unlock-upon-failure?expand=1#diff-5c67a272b20312f1455516764b31abd1R302):
image

This is how it looks in the old config edit (rails):

This is how it looks in the new config edit (spa):

With tooltip:

### Changes in API:

Pipeline config API v5 has been created. Instead of: `"enable_pipeline_locking": true` or `"enable_pipeline_locking": false`, you need to use: `lockBehavior` with one of these three values:
- `lockOnFailure` (same as the old `true`)
- `unlockWhenFinished` (this is new) or
- `none` (same as the old `false`).

### Changes in config XML:

This:
```
<pipeline name="pipeline1" isLocked="true">
  ...
</pipeline>
```

is automatically migrated to:
```
<pipeline name="pipeline1" lockBehavior="lockOnFailure">
  ...
</pipeline>
```

Also, this:
```
<pipeline name="pipeline1" isLocked="false">
  ...
</pipeline>
```

is automatically migrated to:
```
<pipeline name="pipeline1" lockBehavior="none">
  ...
</pipeline>
```

The other new option is: `lockBehavior="unlockWhenFinished"`.

### CODE REVIEW

**Comment by arvindsv** [comment]
Checked with this config:

```
<pipeline name="test">
  <materials>
    <git url="/tmp/test.git" />
  </materials>
  <stage name="stage1">
    <jobs>
      <job name="stage1_job1">
        <tasks>
          <exec command="bash">
            <arg>-c</arg>
            <arg>while [ ! -f "/tmp/p/${GO_PIPELINE_LABEL}/1" ]; do sleep 1; date; done; exit $(cat "/tmp/p/${GO_PIPELINE_LABEL}/1")</arg>
            <runif status="passed" />
          </exec>
        </tasks>
      </job>
    </jobs>
  </stage>
  <stage name="stage2">
    <jobs>
      <job name="stage2_job1">
        <tasks>
          <exec command="bash">
            <arg>-c</arg>
            <arg>while [ ! -f "/tmp/p/${GO_PIPELINE_LABEL}/2" ]; do sleep 1; date; done; exit $(cat "/tmp/p/${GO_PIPELINE_LABEL}/2")</arg>
            <runif status="passed" />
          </exec>
        </tasks>
      </job>
    </jobs>
  </stage>
  <stage name="stage3">
    <approval type="manual" />
    <jobs>
      <job name="stage3_job1">
        <tasks>
          <exec command="bash">
            <arg>-c</arg>
            <arg>while [ ! -f "/tmp/p/${GO_PIPELINE_LABEL}/3" ]; do sleep 1; date; done; exit $(cat "/tmp/p/${GO_PIPELINE_LABEL}/3")</arg>
            <runif status="passed" />
          </exec>
        </tasks>
      </job>
    </jobs>
  </stage>
</pipeline>
```

and controlled it with:

```
LABEL=2; STAGE=1; STATUS=0; mkdir -p "/tmp/p/${LABEL}"; echo "${STATUS}" > "/tmp/p/${LABEL}"/"${STAGE}"
```

**Comment by arvindsv** [comment]
@rajiesh The only part that is left in this is writing a new functional test for this behavior. Should I try and write a new one, or do you think we can add this behavior to an existing test?

**Comment by rajiesh** [comment]
@arvindsv We can write new specs for this. I Can work on it on Monday

File: config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java [hunk 1]
```diff
@@ -231,6 +235,13 @@ private boolean validateLabelTemplateTruncation(String labelTemplate) {
         return LABEL_TEMPATE_ZERO_TRUNC_BLOCK_PATTERN.matcher(labelTemplate).find();
     }
 
+    private void validateLockBehaviorValues() {
```
[the remark this answers is not in the document]
Comment (arvindsv) [reply config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java]: Done.

File: plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/configrepo/ConfigRepoMigrator.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+public class ConfigRepoMigrator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(JsonMessageHandler1_0.class);
+    private final Map<Integer, String> targetVersionToJoltTransformFile;
+
+    public ConfigRepoMigrator() {
+        targetVersionToJoltTransformFile = new HashMap<>();
+        targetVersionToJoltTransformFile.put(2, "/v1_to_v2_enable_pipeline_locking_to_lock_behavior.json");
```
[the remark this answers is not in the document]
Comment (arvindsv) [reply plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/configrepo/ConfigRepoMigrator.java]: Done.

File: plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/configrepo/JsonMessageHandler1_0.java [hunk 3]
```diff
@@ -73,7 +75,7 @@ public CRParseResult responseMessageForParseDirectory(String responseBody) {
                 int version = responseMap.target_version;
 
                 while (version < CURRENT_CONTRACT_VERSION) {
-                    migrate(responseBody, version);
+                    responseBody = migrate(responseBody, version + 1);
```
[the remark this answers is not in the document]
Comment (arvindsv) [reply plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/configrepo/JsonMessageHandler1_0.java]: Done.

**Comment by arvindsv** [comment]
@jyotisingh If you have suggestions for `unlockWhenFinished` and `lockOnFailure`, I can change them. Will need to change in the plugins, functional tests and docs PRs as well, but I can. Let me know what sounds better.

`lockBehavior=...`

File: config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (arvindsv) [reply config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java]: I've left it in the `PipelineConfig` class as well. Feels more correct.

[the remark this answers is not in the document]
Comment (arvindsv) [reply config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java]: I mean, it's now in both places, `CRPipeline` and `PipelineConfig`.

**Comment by arvindsv** [comment]
About that case (v4 and v5), it's expected that you will overwrite it. That's why you shouldn't use older versions. :) There's no way (without changing v4's code) that this can be handled, right?

File: config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jyotisingh) [reply config/config-api/src/com/thoughtworks/go/config/PipelineConfig.java]: Yes, that's exactly what I was asking for.

**Comment by jyotisingh** [comment]
Regarding names - No, I don't have a better suggestion in mind. Thought of making you come up with one, but if you can't think of anything else, then let's stick to it. 

Regarding v4 &v5 - I can't think of a way to get around the problem so thought may be we should call it out. Not sure if release notes is the right place or GitHub issue - of course it would be a "won't fix" with reasons. But we wouldn't be able to find this discussion between us when someone reports it at a later point.

**Comment by arvindsv** [comment]
Ok, I opened and closed https://github.com/gocd/gocd/issues/3958 for this.

**Comment by jyotisingh** [comment]
Cool, PR looks good to me. Run the PR pipeline. Will merge it when the build passes.

**Comment by arvindsv** [comment]
The pipeline seems ok. The functional tests need [my PR](https://github.com/gocd/functional-tests/pull/229) to work properly.

**Comment by arvindsv** [comment]
@jyotisingh - At some point, do you mind merging these three PRs?

https://github.com/gocd/api.go.cd/pull/213
https://github.com/gocd/docs.go.cd/pull/122
https://github.com/gocd/plugin-api.go.cd/pull/44

Not very urgent.

**Pull request merged**